mqtt/mqtt-bridge/src/client.rs

```diff
@@ -0,0 +1,227 @@
+                        None
+                    }
+                }
+            };
+
+            let io = TcpStream::connect(address).await;
+            io.map(|io| (io, password))
+        })
+    }
+}
+
+pub struct MqttClient<T>
+where
+    T: EventHandler,
+{
+    client: Client<TcpConnection<SasTokenSource>>,
+    event_handler: T,
+}
+
+impl<T: EventHandler> MqttClient<T> {
+    pub fn new(
+        address: &str,
+        keep_alive: Duration,
+        clean_session: bool,
+        event_handler: T,
+        connection_credentials: &Credentials,
+    ) -> Self {
+        let (client_id, token_source) = match connection_credentials {
+            Credentials::Provider(provider_settings) => (
+                provider_settings.device_id().into(),
+                Some(SasTokenSource::new(connection_credentials.clone())),
+            ),
+            Credentials::PlainText(creds) => (
+                creds.client_id().into(),
+                Some(SasTokenSource::new(connection_credentials.clone())),
+            ),
+            Credentials::Anonymous(client_id) => (client_id.into(), None),
+        };
+
+        let client = if clean_session {
+            Client::new(
+                Some(client_id),
+                None,
+                None,
+                TcpConnection {
+                    address: address.into(),
+                    token_source,
+                },
+                DEFAULT_MAX_RECONNECT,
+                keep_alive,
+            )
+        } else {
+            Client::from_state(
+                client_id,
+                None,
+                None,
+                TcpConnection {
+                    address: address.into(),
+                    token_source,
+                },
+                DEFAULT_MAX_RECONNECT,
+                keep_alive,
+            )
+        };
+
+        MqttClient {
+            client,
+            event_handler,
+        }
+    }
+
+    pub fn shutdown_handle(&self) -> Result<ShutdownHandle, ShutdownError> {
+        self.client
+            .shutdown_handle()
+            .map_or(Err(ShutdownError::ClientDoesNotExist), |shutdown_handle| {
+                Ok(ShutdownHandle(shutdown_handle))
+            })
+    }
+
+    pub async fn handle_events(mut self) -> Result<(), ClientConnectError> {
+        while let Some(event) = self.client.try_next().await.unwrap_or_else(|e| {
+            error!(message = "failed to poll events", error=%e);
+            None
+        }) {
+            if let Err(e) = self.event_handler.handle_event(event).await {
+                error!("error processing event {}", e);
+            }
+        }
+
+        Ok(())
+    }
+
+    pub async fn subscribe(&mut self, topics: Vec<String>) -> Result<(), ClientConnectError> {
+        debug!("subscribing to topics");
+        let subscriptions = topics.iter().map(|topic| proto::SubscribeTo {
+            topic_filter: topic.to_string(),
+            qos: proto::QoS::AtLeastOnce,
```

Comment:
shouldn't we take QoS from outside? while it is QoS1 for iothub topics, it can be QoS0 for everything else

Reply:
I added a default qos for now. When we talked offline I said that I'll add it to topic, but it's easier to have it as const and change it when we have a final way for settings